colibre-zooms/concatenate_catalogues.py

```diff
@@ -0,0 +1,180 @@
+
+
+def concatenate_catalogues(paths: List[str], filename: str):
+    """
+    Performs concatenation. Saves the concatenated data into a new file
+    Parameters
+    ----------
+    paths: str
+    Path to halo catalogues to be concatenated.
+
+    filename: str
+    Name of the output file
+
+    Returns
+    -------
+    """
+
+    N_zooms = len(paths)
+
+    if N_zooms > 0:
+        concatenated_data = h5.File(filename, "w")
+    else:
+        raise IOError("No data were provided for concatenation!!!")
+
+    # Keep record which files are concatenated
+    concatenated_data.create_group("ConcatenatedFiles")
+
+    for count, path in enumerate(paths, start=1):
+        print(f"\n Opening zoom catalogue {count}...")
+        with h5.File(path, "r") as data:
+
+            # Append 'path' to keep record
+            concatenated_data["ConcatenatedFiles"].attrs.create(
+                path, ""
+            )
+
+            # Copy global attributes
+            if count == 1:
+                for glb_attr in list(data.attrs):
+                    concatenated_data.attrs.create(glb_attr, data.attrs[glb_attr])
+
+            # Loop over fields
+            for key in list(data.keys()):
+                # Create new fields based on the data from the first processed catalogue
+                if count == 1:
+                    try:
+                        # Chunks = True and max shape = None enable resize method
+                        concatenated_data.create_dataset(
+                            key,
+                            data=data[key][:],
+                            shape=data[key].shape,
+                            chunks=True,
+                            dtype=data[key].dtype,
+                            maxshape=(None,),
+                        )
+
+                    # Groups like 'SimulationInfo' have only attributes and no values
+                    except AttributeError:
+                        print(
+                            f"Unable to fetch data for key '{key}'. Creating group... "
+                        )
+                        concatenated_data.create_group(key)
+
+                    # Copy attributes for each field
+                    for attr in list(data[key].attrs):
+                        concatenated_data[key].attrs.create(attr, data[key].attrs[attr])
+
+                # The field has already been created. Use resize to append new data
+                else:
+                    try:
+                        # Fetch fields
+                        concatenated_field = concatenated_data[key]
+                        new_field = data[key]
+
+                        # Compute new shape
+                        new_shape = (concatenated_field.shape[0] + new_field.shape[0],)
+
+                        # Append new data via resize method
+                        concatenated_field.resize(new_shape)
+                        concatenated_field[-new_field.shape[0] :] = new_field[:]
```

Comment:
You may have some problems with this, as I am not sure you can ever remove a part of a HDF5 file (in python at least), it's just hidden. I would worry that when you resize it creates a new copy and leaves the old stuff dormant. Not something that _needs_ to be changed now, but if filesizes are larger than expected you'll need to combine them in memory and only write once.

Reply:
Yes, that's definitely true! 

Although for halo catalogue files, that will never be a problem because memory-wise these files are negligible compared to snapshots. Next, for snapshots, I already tried to reduce the memory usage a little bit by processing only those fields that we really need.

Of course, if in the future we happen to deal with a much greater number of particles, this issue will have to be revised.